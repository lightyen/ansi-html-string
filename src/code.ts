// reference:
//
// https://ttssh2.osdn.jp/manual/4/en/about/ctrlseq.html
// https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda

export enum ANSI {
	Undefined = Number.NaN,
	NUL = 0x00,
	SOH = 0x01,
	STX = 0x02,
	ETX = 0x03,
	EOT = 0x04,
	ENQ = 0x05,
	ACK = 0x06,
	BEL = 0x07,
	BS = 0x08,
	HT = 0x09,
	LF = 0x0a,
	VT = 0x0b,
	FF = 0x0c,
	CR = 0x0d,
	SO = 0x0e,
	SI = 0x0f,
	DLE = 0x10,
	DC1 = 0x11,
	DC2 = 0x12,
	DC3 = 0x13,
	DC4 = 0x14,
	NAK = 0x15,
	SYN = 0x16,
	ETB = 0x17,
	CAN = 0x18,
	EM = 0x19,
	SUB = 0x1a,
	ESC = 0x1b,
	DCS = 0x90, // ESC P
	CSI = 0x9b, // ESC [
	ST = 0x9c, // ESC \
	OSC = 0x9d, // ESC ]
	PM = 0x9e, // ESC ^
	APC = 0x9f, // ESC _
	Space = 32,
	Exclamation = 33,
	DoubleQuote = 34,
	Hash = 35,
	Dollar = 36,
	Percent = 37,
	And = 38,
	SingleQuote = 39,
	LeftRoundBracket = 40,
	RightRoundBracket = 41,
	Asterisk = 42,
	Plus = 43,
	Comma = 44,
	Hyphen = 45,
	Dot = 46,
	Slash = 47,
	_0 = 48,
	_1,
	_2,
	_3,
	_4,
	_5,
	_6,
	_7,
	_8,
	_9,
	Colon = 58,
	SemiColon = 59,
	LessThan = 60,
	Equal = 61,
	GreaterThan = 62,
	Question = 63,
	At = 64,
	A = 65,
	B,
	C,
	D,
	E,
	F,
	G,
	H,
	I,
	J,
	K,
	L,
	M,
	N,
	O,
	P,
	Q,
	R,
	S,
	T,
	U,
	V,
	W,
	X,
	Y,
	Z,
	LeftSquareBracket = 91,
	Backslash = 92,
	RightSquareBracket = 93,
	Caret = 94,
	Underscore = 95,
	Backstick = 96,
	a = 97,
	b,
	c,
	d,
	e,
	f,
	g,
	h,
	i,
	j,
	k,
	l,
	m,
	n,
	o,
	p,
	q,
	r,
	s,
	t,
	u,
	v,
	w,
	x,
	y,
	z,
	LeftCurlyBracket = 123,
	VerticalBar = 124,
	RightCurlyBracket = 125,
	Tilde = 126,
}

export enum SGR {
	// reset all attribute
	Reset = 0,
	// set attribute
	Bold = 1,
	Dim = 2,
	Italic = 3,
	Underline = 4,
	SlowBlink = 5,
	RapidBlink = 6,
	Inverse = 7,
	Hidden = 8,
	Strike = 9,
	// reset attribute
	NotBold = 21,
	NotDim = 22,
	NotItalic = 23,
	NotUnderline = 24,
	NotBlinking = 25,
	NotInverse = 27,
	NotHidden = 28,
	NotStrike = 29,
	// set foreground color
	FgBlack = 30,
	FgRed = 31,
	FgGreen = 32,
	FgYellow = 33,
	FgBlue = 34,
	FgMagenta = 35,
	FgCyan = 36,
	FgWhite = 37,
	FgExt = 38,
	FgReset = 39,
	// set background color
	BgBlack = 40,
	BgRed = 41,
	BgGreen = 42,
	BgYellow = 43,
	BgBlue = 44,
	BgMagenta = 45,
	BgCyan = 46,
	BgWhite = 47,
	BgExt = 48,
	BgReset = 49,
	// set bright foreground color
	BrightFgGray = 90,
	BrightFgRed = 91,
	BrightFgGreen = 92,
	BrightFgYellow = 93,
	BrightFgBlue = 94,
	BrightFgMagenta = 95,
	BrightFgCyan = 96,
	BrightFgWhite = 97,
	// set bright background color
	BrightBgGray = 100,
	BrightBgRed = 101,
	BrightBgGreen = 102,
	BrightBgYellow = 103,
	BrightBgBlue = 104,
	BrightBgMagenta = 105,
	BrightBgCyan = 106,
	BrightBgWhite = 107,
}

export function isNumber(charCode: number) {
	return charCode >= 48 && charCode < 58
}

export function isPrintable(charCode: number) {
	return (charCode >= 32 && charCode < 127) || charCode >= 256
}
