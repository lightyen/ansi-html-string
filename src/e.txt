[SOFT] ! ; : < = > ? - + * & $ % # " ' ( ) , . / ` [digit] [space]
[HARD] ~ @ ^ _ | [ ] { } [alpha]
