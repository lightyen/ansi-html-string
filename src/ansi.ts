import { ASCII, isNumber, SGR } from "./code"
import { ColorMode, ContrastCache, createPalette, ensureContrastRatio, ThemeConfig, toCss, toRgb } from "./colors"

export interface Attributes {
	fgIndexOrRgb: number
	bgIndexOrRgb: number
	fgMode: ColorMode
	bgMode: ColorMode
	bold: boolean
	dim: boolean
	underline: boolean
	inverse: boolean
	italic: boolean
	strike: boolean
	hidden: boolean
}

export interface Word {
	fgMode: ColorMode
	foreground?: string
	bgMode: ColorMode
	background?: string
	bold: boolean
	dim: boolean
	underline: boolean
	italic: boolean
	strike: boolean
	hidden: boolean
	value: string
}

export interface AnchorWord {
	params: Record<string, string>
	url: string
	words: Word[]
}

export function isAnchorWord(w: Word | AnchorWord | undefined): w is AnchorWord {
	return Object.prototype.hasOwnProperty.call(w, "words")
}

export interface Options {
	/** minimum contrast ratio: 1 - 21 (default: 3) */
	minimumContrastRatio?: number
	/** mode (default: inline) */
	mode?: "inline" | "class"
	/** user theme */
	theme?: ThemeConfig
}

export const defaultMinimumContrastRatio = 3

export interface Context extends Attributes {
	contrastCache: ContrastCache
	minimumContrastRatio: number
	mode: "inline" | "class"
	palette: ReturnType<typeof createPalette>
}

export function createContext({
	minimumContrastRatio = defaultMinimumContrastRatio,
	mode = "inline",
	theme,
}: Options = {}): Context {
	return {
		contrastCache: new ContrastCache(),
		minimumContrastRatio,
		mode,
		palette: createPalette(theme),
		fgIndexOrRgb: -1,
		bgIndexOrRgb: -1,
		fgMode: ColorMode.DEFAULT,
		bgMode: ColorMode.DEFAULT,
		bold: false,
		dim: false,
		underline: false,
		inverse: false,
		italic: false,
		strike: false,
		hidden: false,
	}
}

export function parseWithContext(ctx: Context, rawText: string) {
	const _words: Array<Word | AnchorWord> = []
	let a = 0
	let b = 0
	let words = _words
	let anchor: AnchorWord | undefined
	while (b < rawText.length) {
		const char = rawText.charCodeAt(b)
		if (char === ASCII.ESC) {
			pushWord()
			const nextChar = rawText.charCodeAt(b + 1)
			if (nextChar === ASCII.LeftSquareBracket) {
				b = readCSI(b + 2)
			} else if (nextChar === ASCII.RightSquareBracket) {
				b = readOSC(b + 2)
			} else {
				b = other(b + 1)
			}
			a = b
			continue
		} else if (char === ASCII.CSI) {
			pushWord()
			b = readCSI(b + 1)
			a = b
			continue
		} else if (char === ASCII.OSC) {
			pushWord()
			b = readOSC(b + 1)
			a = b
			continue
		} else {
			switch (char) {
				case ASCII.BS:
				case ASCII.DEL: {
					pushWord()
					a = b
					let lastWord = words.at(words.length - 1)
					lastWord = isAnchorWord(lastWord) ? lastWord.words[lastWord.words.length - 1] : lastWord
					if (lastWord) lastWord.value = lastWord.value.slice(0, -1)
					break
				}
				case ASCII.NUL:
				case ASCII.HT:
				case ASCII.LF:
					break
				default:
					if (char < 0x1f || char === ASCII.DEL) {
						pushWord()
						a = b + 1
					}
					break
			}
		}

		b++
	}
	pushWord()
	if (anchor) _words.push(anchor)
	return _words

	function pushWord() {
		if (a < b) words.push(makeWord(rawText.slice(a, b)))
	}

	function makeWord(value: string): Word {
		let fgColor = ctx.fgIndexOrRgb
		let bgColor = ctx.bgIndexOrRgb
		let fgMode = ctx.fgMode
		let bgMode = ctx.bgMode

		if (ctx.inverse) {
			let tmp = fgColor
			fgColor = bgColor
			bgColor = tmp
			tmp = fgMode
			fgMode = bgMode
			bgMode = tmp
		}

		if (fgMode === ColorMode.P256 && fgColor < 16) fgMode = ColorMode.P16
		if (bgMode === ColorMode.P256 && bgColor < 16) bgMode = ColorMode.P16

		const foreground =
			ctx.mode === "class" && fgMode !== ColorMode.RGB
				? getForegroundClass(fgMode, fgColor, ctx.inverse, ctx.bold)
				: getForegroundCss(bgMode, bgColor, fgMode, fgColor, ctx.inverse, ctx.bold)

		const background =
			ctx.mode === "class" && bgMode !== ColorMode.RGB
				? getBackgroundClass(bgMode, bgColor, ctx.inverse)
				: getBackgroundCss(bgMode, bgColor, ctx.inverse)

		const word: Word = {
			fgMode,
			foreground,
			bgMode,
			background,
			bold: ctx.bold,
			dim: ctx.dim,
			underline: ctx.underline,
			italic: ctx.italic,
			strike: ctx.strike,
			hidden: ctx.hidden,
			value,
		}

		return word
	}

	function getForegroundCss(
		bgColorMode: ColorMode,
		bgIndexOrRgb: number,
		fgColorMode: ColorMode,
		fgIndexOrRgb: number,
		inverse: boolean,
		bold: boolean,
	): string | undefined | never {
		const minimumContrastCss = getMinimumContrastCss(
			bgColorMode,
			bgIndexOrRgb,
			fgColorMode,
			fgIndexOrRgb,
			inverse,
			bold,
		)
		if (minimumContrastCss) return minimumContrastCss
		return _getForegroundCss(fgColorMode, fgIndexOrRgb, inverse, bold)
	}

	function getBackgroundCss(
		bgColorMode: ColorMode,
		bgIndexOrRgb: number,
		inverse: boolean,
	): string | undefined | never {
		switch (bgColorMode) {
			case ColorMode.P16:
			case ColorMode.P256: {
				const color = ctx.palette.colors.at(bgIndexOrRgb)
				if (!color) throw Error(`background is not defined: ${bgIndexOrRgb}`)
				return color.css
			}
			case ColorMode.RGB:
				return toCss(bgIndexOrRgb)
			case ColorMode.DEFAULT:
			default:
				if (inverse) return ctx.palette.foreground?.css
				return ctx.palette.background?.css
		}
	}

	function getForegroundClass(
		fgColorMode: ColorMode,
		fgIndexOrRgb: number,
		inverse: boolean,
		bold: boolean,
	): string | undefined | never {
		switch (fgColorMode) {
			case ColorMode.P16:
			case ColorMode.P256:
				if (bold && fgIndexOrRgb < 8) fgIndexOrRgb += 8
				return `${fgIndexOrRgb}`
			case ColorMode.DEFAULT:
			default:
				if (inverse) return "inverse"
				return undefined
		}
	}

	function getBackgroundClass(
		bgColorMode: ColorMode,
		bgIndexOrRgb: number,
		inverse: boolean,
	): string | undefined | never {
		switch (bgColorMode) {
			case ColorMode.P16:
			case ColorMode.P256:
				return `${bgIndexOrRgb}`
			case ColorMode.DEFAULT:
			default:
				if (inverse) return "inverse"
				return undefined
		}
	}

	function getMinimumContrastCss(
		bgColorMode: ColorMode,
		bgIndexOrRgb: number,
		fgColorMode: ColorMode,
		fgIndexOrRgb: number,
		inverse: boolean,
		bold: boolean,
	) {
		if (ctx.minimumContrastRatio <= 1) return undefined
		const bg = (bgIndexOrRgb << 8) | bgColorMode
		const fg = (fgIndexOrRgb << 8) | fgColorMode
		const adjustedColor = ctx.contrastCache.getCss(bg, fg)
		if (adjustedColor !== undefined) return adjustedColor || undefined

		const fgRgb = getForegroundRgb(fgColorMode, fgIndexOrRgb, inverse, bold)
		const bgRgb = getBackgroundRgb(bgColorMode, bgIndexOrRgb, inverse)

		if (fgRgb != undefined && bgRgb != undefined) {
			const rgb = ensureContrastRatio(fgRgb, bgRgb, ctx.minimumContrastRatio)
			if (!rgb) {
				ctx.contrastCache.setCss(bg, fg, null)
				return undefined
			}

			const css = toCss(rgb)
			ctx.contrastCache.setCss(bg, fg, css)
			return css
		}

		return undefined
	}

	function _getForegroundCss(
		fgColorMode: ColorMode,
		fgIndexOrRgb: number,
		inverse: boolean,
		bold: boolean,
	): string | undefined | never {
		switch (fgColorMode) {
			case ColorMode.P16:
			case ColorMode.P256: {
				if (bold && fgIndexOrRgb < 8) fgIndexOrRgb += 8
				const color = ctx.palette.colors.at(fgIndexOrRgb)
				if (!color) throw Error(`foreground is not defined: ${fgIndexOrRgb}`)
				return color.css
			}
			case ColorMode.RGB:
				return toCss(fgIndexOrRgb)
			case ColorMode.DEFAULT:
			default:
				if (inverse) return ctx.palette.background?.css
				return ctx.palette.foreground?.css
		}
	}

	function getForegroundRgb(
		fgColorMode: ColorMode,
		fgIndexOrRgb: number,
		inverse: boolean,
		bold: boolean,
	): number | undefined | never {
		switch (fgColorMode) {
			case ColorMode.P16:
			case ColorMode.P256: {
				if (bold && fgIndexOrRgb < 8) fgIndexOrRgb += 8
				const color = ctx.palette.colors.at(fgIndexOrRgb)
				if (!color) throw Error(`foreground is not defined: ${fgIndexOrRgb}`)
				return color.rgb
			}
			case ColorMode.RGB:
				return fgIndexOrRgb
			case ColorMode.DEFAULT:
			default:
				if (inverse) return ctx.palette.background?.rgb
				return ctx.palette.foreground?.rgb
		}
	}

	function getBackgroundRgb(bgColorMode: number, bgIndexOrRgb: number, inverse: boolean): number | undefined | never {
		switch (bgColorMode) {
			case ColorMode.P16:
			case ColorMode.P256: {
				const color = ctx.palette.colors.at(bgIndexOrRgb)
				if (!color) throw Error(`background is not defined: ${bgIndexOrRgb}`)
				return color.rgb
			}
			case ColorMode.RGB:
				return bgIndexOrRgb
			case ColorMode.DEFAULT:
			default:
				if (inverse) return ctx.palette.foreground?.rgb
				return ctx.palette.background?.rgb
		}
	}

	function getNumber(a: number, b: number): number {
		let ans = 0
		for (let k = a; k < b; k++) {
			ans = 10 * ans + rawText.charCodeAt(k) - ASCII._0
		}
		return ans
	}

	/** @return last index */
	function readCSI(index: number): number {
		if (index >= rawText.length) return index
		let b = index

		while (b < rawText.length) {
			const char = rawText.charCodeAt(b)
			if (isEnd(char)) {
				handle(index, b + 1)
				b++
				break
			}
			b++
		}
		return b

		function isEnd(char: number) {
			return char < 0x20 || char >= 0x40
		}

		function handle(a: number, b: number): void {
			b--
			const lastChar = rawText.charCodeAt(b)
			switch (lastChar) {
				case ASCII.m:
					{
						const attrs: number[] = []
						let i = a
						while (i < b) {
							const c = rawText.charCodeAt(i)
							if (c === ASCII.SemiColon) {
								attrs.push(getNumber(a, i))
								i++
								a = i
							} else if (isNumber(c)) {
								i++
							} else {
								// failed
								return
							}
						}
						attrs.push(getNumber(a, i))
						setAttributes(attrs)
					}
					break
				case ASCII.J:
					{
						if (rawText.charCodeAt(a) === ASCII.Question) a++
						let i = a
						while (i < b) {
							const c = rawText.charCodeAt(i)
							if (isNumber(c)) {
								i++
							} else {
								// failed
								return
							}
						}
						const v = getNumber(a, i)
						if (v === 1 || v === 2) words.splice(0)
					}

					break
			}
		}

		function resetAttributes() {
			ctx.fgIndexOrRgb = -1
			ctx.bgIndexOrRgb = -1
			ctx.fgMode = ColorMode.DEFAULT
			ctx.bgMode = ColorMode.DEFAULT
			ctx.bold = false
			ctx.dim = false
			ctx.italic = false
			ctx.underline = false
			ctx.inverse = false
			ctx.hidden = false
			ctx.strike = false
		}

		function setAttributes(attrs: number[]) {
			for (let i = 0; i < attrs.length; i++) {
				const code = attrs[i]
				switch (code) {
					case SGR.Reset:
						resetAttributes()
						break
					case SGR.Bold:
						ctx.bold = true
						break
					case SGR.Dim:
						ctx.dim = true
						break
					case SGR.Italic:
						ctx.italic = true
						break
					case SGR.Underline:
						ctx.underline = true
						break
					case SGR.Inverse:
						ctx.inverse = true
						break
					case SGR.Hidden:
						ctx.hidden = true
						break
					case SGR.Strike:
						ctx.strike = true
						break
					case SGR.SlowBlink:
					case SGR.RapidBlink:
						// undefined
						break
				}
				if (code >= SGR.FgBlack && code <= SGR.FgWhite) {
					ctx.fgIndexOrRgb = code - SGR.FgBlack
					ctx.fgMode = ColorMode.P16
				} else if (code >= SGR.BgBlack && code <= SGR.BgWhite) {
					ctx.bgIndexOrRgb = code - SGR.BgBlack
					ctx.bgMode = ColorMode.P16
				} else if (code >= SGR.BrightFgGray && code <= SGR.BrightFgWhite) {
					ctx.fgIndexOrRgb = 8 + code - SGR.BrightFgGray
					ctx.fgMode = ColorMode.P16
				} else if (code >= SGR.BrightBgGray && code <= SGR.BrightBgWhite) {
					ctx.bgIndexOrRgb = 8 + code - SGR.BrightBgGray
					ctx.bgMode = ColorMode.P16
				} else if (code === SGR.FgReset) {
					ctx.fgIndexOrRgb = -1
					ctx.fgMode = ColorMode.DEFAULT
				} else if (code === SGR.BgReset) {
					ctx.bgIndexOrRgb = -1
					ctx.bgMode = ColorMode.DEFAULT
				} else if (code === SGR.FgExt) {
					if (attrs[i + 1] === 5) {
						ctx.fgMode = ColorMode.P256
						if (i + 2 >= attrs.length) {
							ctx.fgIndexOrRgb = 0
							break
						}
						ctx.fgIndexOrRgb = attrs[i + 2]
						i += 2
					} else if (attrs[i + 1] === 2) {
						ctx.fgMode = ColorMode.RGB
						if (i + 4 >= attrs.length) {
							ctx.fgIndexOrRgb = 0
							break
						}
						const r = attrs[i + 2]
						const g = attrs[i + 3]
						const b = attrs[i + 4]
						ctx.fgIndexOrRgb = toRgb(r, g, b)

						i += 4
					}
				} else if (code === SGR.BgExt) {
					if (attrs[i + 1] === 5) {
						ctx.bgMode = ColorMode.P256
						if (i + 2 >= attrs.length) {
							ctx.bgIndexOrRgb = 0
							break
						}
						ctx.bgIndexOrRgb = attrs[i + 2]
						i += 2
					} else if (attrs[i + 1] === 2) {
						ctx.bgMode = ColorMode.RGB
						if (i + 4 >= attrs.length) {
							ctx.bgIndexOrRgb = 0
							break
						}
						const r = attrs[i + 2]
						const g = attrs[i + 3]
						const b = attrs[i + 4]
						ctx.bgIndexOrRgb = toRgb(r, g, b)
						i += 4
					}
				}
			}
		}
	}

	/** @return last index */
	function readOSC(index: number): number {
		// format: OSC 8 ; params ; URI ST
		// params format: id=xyz123:foo=bar:baz=quux
		if (index >= rawText.length) return index
		let mode: number = Number.NaN
		let state = 0
		let params: Record<string, string> = {}
		let a = index
		let b = index
		while (b < rawText.length) {
			const char = rawText.charCodeAt(b)
			if (char === ASCII.ST || char === ASCII.BEL) {
				addAnchor()
				b = b + 1
				break
			}
			if (char === ASCII.ESC && rawText.charCodeAt(b + 1) === ASCII.Backslash) {
				addAnchor()
				b = b + 2
				break
			}
			if (char !== ASCII.SemiColon) {
				b++
				continue
			}
			if (Number.isNaN(mode)) {
				mode = getNumber(a, b)
				b++
				a = b
				continue
			}

			if (mode === 8) {
				if (state === 0) {
					params = getMap(rawText.slice(a, b))
					state = 1
					b++
					a = b
					continue
				}

				// failed
				state = 2

				b++
				a = b
				continue
			}

			b++
		}

		return b

		function addAnchor() {
			if (anchor) {
				_words.push(anchor)
				anchor = undefined
				words = _words
			}
			if (state === 1 && a < b) {
				const url = rawText.slice(a, b)
				anchor = { url, params, words: [] }
				words = anchor.words
			}
		}

		function getMap(value: string) {
			const result: Record<string, string> = {}
			value.split(":").forEach(val => {
				const i = val.indexOf("=")
				if (i > 0) {
					result[val.slice(0, i)] = val.slice(i + 1)
				}
			})
			return result
		}
	}

	/** @return last index */
	function other(index: number): number {
		if (index >= rawText.length) return index
		let b = index
		while (b < rawText.length) {
			const char = rawText.charCodeAt(b)
			if (isEnd(char)) {
				b++
				break
			}
			b++
		}
		return b

		function isEnd(char: number) {
			return true
		}
	}
}
