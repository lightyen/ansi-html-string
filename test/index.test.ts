import { toDemo } from "../src"

it("render", () => {
	const text = `MacroError:
src/AppMain.tsx

\x1b[38;2;255;131;131m&#10005; \x1b[38;2;255;211;211mbg-indigoa-400\x1b[39m\x1b[38;2;255;131;131m was not found\x1b[39m

Try one of these classes:

\x1b[38;2;153;153;153m-\x1b[39m \x1b[93mbg-indigo-400\x1b[39m \x1b[38;2;153;153;153m&gt;\x1b[39m \x1b[38;2;129;140;248m&#9635; \x1b[39m#818cf8
\x1b[38;2;153;153;153m-\x1b[39m \x1b[93mbg-indigo-100\x1b[39m \x1b[38;2;153;153;153m&gt;\x1b[39m \x1b[38;2;224;231;255m&#9635; \x1b[39m#e0e7ff
\x1b[38;2;153;153;153m-\x1b[39m \x1b[93mbg-indigo-200\x1b[39m \x1b[38;2;153;153;153m&gt;\x1b[39m \x1b[38;2;199;210;254m&#9635; \x1b[39m#c7d2fe
\x1b[38;2;153;153;153m-\x1b[39m \x1b[93mbg-indigo-300\x1b[39m \x1b[38;2;153;153;153m&gt;\x1b[39m \x1b[38;2;165;180;252m&#9635; \x1b[39m#a5b4fc
\x1b[38;2;153;153;153m-\x1b[39m \x1b[93mbg-indigo-500\x1b[39m \x1b[38;2;153;153;153m&gt;\x1b[39m \x1b[38;2;99;102;241m&#9635; \x1b[39m#6366f1

	Standard colors:
	\x1b[48;5;0m  \x1b[48;5;1m  \x1b[48;5;2m  \x1b[48;5;3m  \x1b[48;5;4m  \x1b[48;5;5m  \x1b[48;5;6m  \x1b[48;5;7m  \x1b[0m
	\x1b[48;5;8m  \x1b[48;5;9m  \x1b[48;5;10m  \x1b[48;5;11m  \x1b[48;5;12m  \x1b[48;5;13m  \x1b[48;5;14m  \x1b[48;5;15m  \x1b[0m

`

	const output = toDemo(text, {
		theme: {
			// background: "#23272e",
			// foreground: "#abb2bf",
			black: "#000000",
			red: "#D34F56",
			green: "#B9C954",
			yellow: "#E6C452",
			blue: "#7CA7D8",
			magenta: "#C299D6",
			cyan: "#73BFB1",
			white: "#FFFFFF",
		},
	})

	expect(output).toMatchSnapshot()
})
