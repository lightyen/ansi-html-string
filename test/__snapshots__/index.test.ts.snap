// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
	<head>
		<meta charset=\\"UTF-8\\">
		<meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
		<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
		<title>Document</title>
	</head>
	<body style=\\"background: undefined\\">
		<pre style=\\"font-size: initial\\">MacroError:
src/AppMain.tsx

<span style=\\"color: #ff8383\\">&#10005; </span><span style=\\"color: #ffd3d3\\">bg-indigoa-400</span><span style=\\"color: #ff8383\\"> was not found</span>

Try one of these classes:

<span style=\\"color: #999999\\">-</span> <span style=\\"color: #f0a45d\\">bg-indigo-400</span> <span style=\\"color: #999999\\">&gt;</span> <span style=\\"color: #818cf8\\">&#9635; </span>#818cf8
<span style=\\"color: #999999\\">-</span> <span style=\\"color: #f0a45d\\">bg-indigo-100</span> <span style=\\"color: #999999\\">&gt;</span> <span style=\\"color: #e0e7ff\\">&#9635; </span>#e0e7ff
<span style=\\"color: #999999\\">-</span> <span style=\\"color: #f0a45d\\">bg-indigo-200</span> <span style=\\"color: #999999\\">&gt;</span> <span style=\\"color: #c7d2fe\\">&#9635; </span>#c7d2fe
<span style=\\"color: #999999\\">-</span> <span style=\\"color: #f0a45d\\">bg-indigo-300</span> <span style=\\"color: #999999\\">&gt;</span> <span style=\\"color: #a5b4fc\\">&#9635; </span>#a5b4fc
<span style=\\"color: #999999\\">-</span> <span style=\\"color: #f0a45d\\">bg-indigo-500</span> <span style=\\"color: #999999\\">&gt;</span> <span style=\\"color: #6366f1\\">&#9635; </span>#6366f1

	Standard colors:
	<span style=\\"background-color: #000000\\">  </span><span style=\\"background-color: #D34F56\\">  </span><span style=\\"background-color: #B9C954\\">  </span><span style=\\"background-color: #E6C452\\">  </span><span style=\\"background-color: #7CA7D8\\">  </span><span style=\\"background-color: #C299D6\\">  </span><span style=\\"background-color: #73BFB1\\">  </span><span style=\\"background-color: #FFFFFF\\">  </span>
	<span style=\\"background-color: #4f5666\\">  </span><span style=\\"background-color: #ff616e\\">  </span><span style=\\"background-color: #a5e075\\">  </span><span style=\\"background-color: #f0a45d\\">  </span><span style=\\"background-color: #4dc4ff\\">  </span><span style=\\"background-color: #de73ff\\">  </span><span style=\\"background-color: #4cd1e0\\">  </span><span style=\\"background-color: #d7dae0\\">  </span>

</pre>
	</body>
</html>"
`;
